### ISSUE TITLE
Quest participants can’t leave the quest

### ISSUE THREAD

**Issue by lemoness** [HabitRPG/habitica#9292]
A quest participant can’t leave the quest (which is necessary if they want to leave the party). The quest owner should NOT be able to leave the quest. The party leader and all other quest participants SHOULD be able to.

EDIT by Alys 2020-04-09: The desired behaviour is to have a feature to allow a player to remove themself from the quest without leaving the party.
The behaviour when leaving a party should not be changed. I.e., if you are in a quest, you cannot leave the party.

**Comment by lemoness** [comment]
Slight addendum to this: the way that we'd prefer to handle this for the short term is to simply make it so that leaving the party is possible during a quest, albeit with a warning when you click the "Leave" button that says "Warning! Leaving the Party now will remove you from your quest and you will not receive any quest rewards. If you are the Quest owner, the Quest will be cancelled."

I'm making this "suggestion-discussion" in case there is additional feedback.

**Comment by Alys** [comment]
It would be essential that leaving the party removed the player from the quest's object and removed the quest data from the player's account. The code that does those two things should include its own check that both had been done successfully, and if not, it should repeat the failed database updates until they were successful. In the past we've had some unpleasant bugs resulting from quest data remaining after a player left the party.

**Comment by Alys** [comment]
That's not to say that I disagree with @lemoness's comment. From the user's point of view, I think it would be fine if the only change was that they got the warning message. However the contributor who implements that warning message should also implement the code to remove the quest data.

I suggest that the final sentence in the warning could be "If you are the Quest owner, the Quest will be aborted and the scroll will be returned to you."  Otherwise we'd probably get repeated questions about the scroll.

**Comment by lemoness** [comment]
Yes, I didn't mean to suggest that there shouldn't be code changes to remove you completely from the quest! That is definitely essential. Thank you for the addendum.

**Comment by veeeeeee** [comment]
Confirm that users should be able to leave a quest while a quest is in progress. This fix will likely need thorough testing to ensure no additional issues are created.

**Comment by Alys** [comment]
There was discussion above about implementing this by having a player removed from the quest when they leave the party.
I'm thinking now that might no longer be the best approach and that it would be better to have a separate feature to allow a player to remove themself from the quest without leaving the party. From the way that @veeeeeee wrote her comment, I'm guessing that's the desired approach now. If there's no objections in the next couple of days, I'll update the top post to indicate that.
A separate feature is more flexible (e.g., there's cases where you'd want to leave a quest but stay in the party), the code and tests will be more encapsulated, and we've found that players don't always read warning messages so allowing leaving a party with a warning that you'll leave the quest too might not work the way we hoped.

**Comment by Alys** [comment]
We already have an API POST route for leaving a Quest: [/api/v3/groups/:groupId/quests/leave](https://habitica.com/apidoc/#api-Quest-LeaveQuest)
So the only change needed for this issue is the website UI for it. For example, there could be a "Leave Quest" button under the Quest box on the Party page, next to the Abort button in the screenshot below. (NB Only the Party leader and Quest owner see the Abort button.)

The "Leave Quest" button (or whatever other UI element is implemented) would need to appear for all Quest members except the Quest owner - the owner should not be able to leave the Quest.

@Tressley what do you think of this idea for the button positioning? I've changed the status label from "help wanted" to "on hold: needs design".

The API route works for only an active quest so there's no need to display a button when the quest is in the invitation stage.

![image](https://user-images.githubusercontent.com/1495809/87777918-b3b20400-c86d-11ea-8f50-37311621ea21.png)

**Comment by Tressley** [comment]
@Alys -- Thank you for this! The Quest column suffers from being overlooked during redesign implementation, but we do have plans to tackle it soon. I have designs that need to be re-worked a little, but this is a use-case the team can discuss before implementation.

**Comment by saraolson** [comment]
this feature is supported now after the quest column was redesigned: participants can leave a quest, quest owner can cancel quest entirely
![Image](https://github.com/user-attachments/assets/f9265110-1c35-46cb-af9b-fcdff178195d)
